README.md
[![Build Status](https://travis-ci.org/yusukebe/Shodo.png?branch=master)](https://travis-ci.org/yusukebe/Shodo)
# NAME

Shodo - Auto-generate documents from HTTP::Request and HTTP::Response

# SYNOPSIS

    use HTTP::Request::Common;
    use HTTP::Response;
    use Shodo;

    my $shodo = Shodo->new();
    my $suzuri = $shodo->new_suzuri('An endpoint method.');

    my $req = POST '/entry', [ id => 1, message => 'Hello Shodo' ];
    $suzuri->request($req);
    my $res = HTTP::Response->new(200);
    $res->content('{ "message" : "success" }');
    $suzuri->response($res);

    print $suzuri->document(); # print document as Markdown format

# DESCRIPTION

Shodo generates Web API documents as Markdown format automatically and validates parameters using HTTP::Request/Response.

__THIS IS A DEVELOPMENT RELEASE. API MAY CHANGE WITHOUT NOTICE__.

# Methods of "Shodo" Module

## new

    my $shodo = Shodo->new(
        document_root => 'doc'
    );

Create and return new Shodo object. "document\_root" is optional parameter for your document root directory.

## template

    $shodo->template($tmpl);

Set custom template.

## document\_root

    $shodo->document_root('doc');

Set document root directory.

## new\_suzuri

    my $suzuri = $shodo->new_suzuri('This is description.');

Create and return new Suzuri object with the description.

## stock

    $shodo->stock($suzuri->doc());

Stock text of documents for writing later. The parameter document is anything ok, but Markdown based is recommended.

## write

    $shodo->write('output.md');

Write the documentation in stocks to the file and make the stock empty.

# Methods of "Suzuri" Module

## request

    $suzuri->request($req);

Set HTTP::Request object.

## response

    $suzuri->response($res);

Set HTTP::Response object.

## params

    $suzuri->params(
        category => { isa => 'Str', documentation => 'Category of articles.' },
        limit => { isa => 'Int', default => 20, optional => 1, documentation => 'Limitation numbers per page.' },
        page => { isa => 'Int', default => 1, optional => 1, documentation => 'Page number you want to get.' }
    );

Parameters for validation and documentation. These rules are based on [Data::Validator](http://search.cpan.org/perldoc?Data::Validator)'s interfaces.

## validate

    $suzuri->validate($params);

Validate with the rules defined by "params" method. Parameter must be HASH ref.

## doc

    $suzuri->doc();

Return the Markdown formatted document for Web API.

# SEE ALSO

"autodoc": [https://github.com/r7kamura/autodoc](https://github.com/r7kamura/autodoc)

[Test::JsonAPI::Autodoc](http://search.cpan.org/perldoc?Test::JsonAPI::Autodoc)

What is Shodo?: [http://en.wikipedia.org/wiki/Shodo](http://en.wikipedia.org/wiki/Shodo)

# THANKS

Songmu for naming as "Shodo". It's pretty.

Moznion for making Test::JsonAPI::Autodoc.

Hachioji.pm for advising.

# LICENSE

Copyright (C) Yusuke Wada.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Yusuke Wada <yusuke@kamawada.com>


Changes
Revision history for Perl extension Shodo

{{$NEXT}}

0.08 2013-12-27T08:17:48Z

    - move Shodo::Test::JSONRPC to "Test::Shodo::JSONRPC"
    - fixed required modules
    - update POD

0.07 2013-11-29T15:48:57Z

    - add '?' to uri if it has query, thanks to hisaichi5518

0.06 2013-11-09T16:25:50Z

    - highlight JSON texts

0.05 2013-11-05T09:18:14Z

    - fixed utf8 bug

0.04 2013-11-05T05:05:22Z

    - no throw on validation error

0.03 2013-11-04T07:30:48Z

    - add required module

0.02 2013-11-04T07:19:56Z

    - update documents
    - add test module "Shodo::Test::JSONRPC"

0.01 2013-11-02T07:26:03Z

    - original version


lib/Shodo/Test/JSONRPC.pm
package Shodo::Test::JSONRPC;
use strict;
use warnings;
use Shodo;
use parent qw/Exporter/;
use JSON qw/from_json/;
use Try::Tiny;
use Test::More;

our @EXPORT = qw/
    shodo_document_root
    shodo_test
    shodo_params
    shodo_req_ok
    shodo_res_ok
    shodo_doc
    shodo_write
/;

my $shodo = Shodo->new();
my $suzuri;

sub shodo_document_root {
    my $dir = shift;
    $shodo->document_root($dir);
}

sub shodo_test {
    my ($description, $coderef) = @_;
    $suzuri = $shodo->new_suzuri($description) unless $suzuri;
    my $result = Test::More::subtest($description => $coderef);
}

sub shodo_params {
    my %args = @_;
    $suzuri->params(%args);
}

sub shodo_req_ok {
    my ($req, $note) = @_;
    $suzuri->request($req);
    my $data = try {
        from_json($req->content);
    }catch{
        warn "$_\n";
    };
    return unless $data;
    return unless $suzuri->validate($data->{params});
    Test::More::ok($req, $note);
}

sub shodo_res_ok {
    my ($res, $code, $note) = @_;
    $suzuri->response($res);
    my $result = Test::More::is $res->code, 200, $note;
    $shodo->stock($suzuri->doc());
    $suzuri = undef;
    return $result;
}

sub shodo_doc {
    return $shodo->stock();
}

sub shodo_write {
    my $filename = shift;
    $shodo->write($filename);
}

1;

__END__

=encoding utf-8

=head1 NAME

Shodo::Test::JSONRPC - Test module using Shodo for JSON-RPC Web API

=head1 SYNOPSIS

    use Test::More;
    use Plack::Test;
    use HTTP::Request;
    use JSON qw/to_json/;
    use Shodo::Test::JSONRPC;

    # PSGI application
    my $app = sub {
        my $data = {
            jsonrpc => '2.0',
            result  => {
                entries =>
                  [ { title => 'Hello', body => 'This is an example.' } ]
            },
            id => 1
        };
        my $json = to_json($data);
        return [ 200, [ 'Content-Type' => 'application/json' ], [$json] ];
    };

    # use Plack::Test
    my $plack_test = Plack::Test->create($app);
    shodo_document_root('sample_documents');

    # shodo_test, like a subtest!
    shodo_test 'get_entries' => sub {
        shodo_params(
            category => { isa => 'Str', documentation => 'Category of articles.' },
            limit => { isa => 'Int', default => 20, optional => 1, documentation => 'Limitation numbers per page.' }
        );
        my $data = {
            jsonrpc => '2.0',
            method  => 'get_entries',
            params  => { limit => 1, category => 'technology' }
        };
        my $json = to_json($data);
        my $req  = HTTP::Request->new(
            'POST', '/',
            [
                'Content-Type'   => 'application/json',
                'Content-Length' => length $json
            ],
            $json
        );
        shodo_req_ok( $req, 'Request is valid!' );
        my $res = $plack_test->request($req);
        shodo_res_ok( $res, 200, 'Response is ok!' ); # auto sock document
    };

    shodo_write('some_methods.md'); # Generate a markdown-formatted document.

    done_testing();

=head1 DESCRIPTION

Shodo-based test module for JSON-RPC Web API.

=head1 LICENSE

Copyright (C) Yusuke Wada.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Yusuke Wada E<lt>yusuke@kamawada.comE<gt>

=cut


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Shodo',
    module_name     => 'Shodo',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Shodo/Hanshi.pm
package Shodo::Hanshi;
use strict;
use warnings;
use Text::MicroTemplate qw/render_mt encoded_string/;

sub new {
    my ($class, %args) = @_;
    my $self = bless { template => $args{template} }, $class;
    $self;
}

sub render {
    my ($self, $stash) = @_;
    my $tmpl = $self->{template} || $self->default_template;
    my $renderer = render_mt($self->header . $tmpl, $stash);
    return $renderer->as_string();
}

sub header {
    my $tmpl = <<'END';
? my ($stash) = @_;
? no strict qw/refs vars/;
? no warnings;
? for my $key (keys %$stash) {
? if (ref $stash->{$key}) {
? ${$key} = $stash->{$key};
? }else{
? ${$key} = encoded_string($stash->{$key});
? }
? }
? use warnings;
? use strict qw/refs subs/;
END
    return $tmpl;
}

sub default_template {
    my $tmpl = <<'END';
## <?= $method ?> <?= $path ?>

? if ($description) {
<?= $description ?>
? }

### Request

```json
<?= $method ?> <?= $path ?><?= $query || '' ?>

<?= $request_body ?>
```
? if ($rule) {

### Parameters

? for my $name (keys %$rule) {
* `<?= $name ?>` - <?= $rule->{$name}{documentation} || '' ?>
? for my $key (keys %{$rule->{$name}}) {
? next if $key eq 'documentation';
? if ($rule->{$name}{$key}) {
  * <?= $key ?>: **<?= $rule->{$name}{$key} || '' ?>**
? }
? }
? }
? }

### Response

```json
Status: <?= $code ?>

<?= $response_body ?>
```

---

END
    return $tmpl;
}

1;


lib/Shodo/Suzuri.pm
package Shodo::Suzuri;
use strict;
use warnings;
use Carp qw//;
use Try::Tiny;
use JSON qw/from_json to_json/;
use Data::Validator;
use Clone qw/clone/;

sub new {
    my ($class, %args) = @_;
    my $self = bless {
        hanshi => $args{hanshi}
    }, $class;
    $self->stash->{description} = $args{description} || '';
    $self;
}

sub hanshi {
    shift->{hanshi};
}

sub stash {
    my $self = shift;
    $self->{stash} ||= {};
    return $self->{stash};
}

sub request {
    my ($self, $req) = @_;
    unless (try { $req->isa('HTTP::Request') }) {
        Carp::croak("Request is not HTTP::Request: $req");
    }
    $self->stash->{method} = $req->method;
    $self->stash->{path} = $req->uri->path;
    $self->stash->{query} = $req->uri->query;
    $self->stash->{request_body} = $req->decoded_content;
    if($req->content_type =~ m!^application/json!) {
        my $json_body = to_json(from_json($self->stash->{request_body}, { utf8 => 1 }), { pretty => 1 });
        $self->stash->{request_body} = $json_body;
    }
    return $req;
}

sub response {
    my ($self, $res) = @_;
    unless (try { $res->isa('HTTP::Response') }) {
        Carp::croak("Response is not HTTP::Response: $res");
    }
    $self->stash->{code} = $res->code;
    $self->stash->{status_line} = $res->status_line;
    $self->stash->{response_body} = $res->decoded_content;
    if($res->content_type =~ m!^application/json!) {
        my $json_body = to_json(from_json($self->stash->{response_body}, { utf8 => 1}), { pretty => 1 });
        $self->stash->{response_body} = $json_body;
    }
    return $res;
}

sub document {
    my $self = shift;
    return $self->hanshi->render( $self->stash );
}

sub params {
    my ($self, %args) = @_;
    $self->stash->{rule} = clone(\%args);
    my $validator = Data::Validator->new( %args )->with('NoThrow');
    $self->{validator} = $validator;
}

sub validate {
    my ($self, @args) = @_;
    Carp::croak "Rule is not set on Suzuri instance" unless $self->{validator};
    my $result;
    if( ref $args[0] && ref $args[0] eq 'HASH') {
        $result = $self->{validator}->validate($args[0]);
    }else{
        $result = $self->{validator}->validate(@args);
    }
    if($self->{validator}->has_errors()) {
        return;
    }
    return $result;
}

*req = \&request;
*res = \&response;
*doc = \&document;
*rule = \&params;

1;


lib/Shodo.pm
package Shodo;
use 5.008005;
use strict;
use warnings;
use Shodo::Suzuri;
use Shodo::Hanshi;
use Path::Tiny qw/path/;

our $VERSION = "0.07";

sub new {
    my ($class, %args) = @_;
    my $self = bless {
        template => $args{template},
        document_root => $args{document_root} ? path($args{document_root}) : path('.'),
    }, $class;
    $self->{stock} = '';
    $self;
}

sub template {
    my ($self, $tmpl) = @_;
    return $self->{template} unless $tmpl;
    $self->{template} = $tmpl;
    return $self->{template};
}

sub document_root {
    my ($self, $path) = @_;
    return $self->{document_root} unless $path;
    $self->{document_root} = path($path);
    return $self->{document_root};
}

sub new_suzuri {
    my ($self, $description) = @_;
    my $hanshi = Shodo::Hanshi->new(
        template => $self->template,
    );
    return Shodo::Suzuri->new(
        hanshi => $hanshi,
        description => $description,
        document_root => $self->document_root
    );
}

sub stock {
    my ($self, $doc) = @_;
    return $self->{stock} unless $doc;
    $self->{stock} .= $doc;
    return $self->{stock};
}

sub write {
    my ($self, $filename) = @_;
    Carp::croak "Document root is not direcotry: " . $self->document_root unless( -d $self->document_root );
    my $file = $self->document_root->child($filename);
    $file->spew_utf8( $self->stock );
    $self->{stock} = '';
}

1;
__END__

=encoding utf-8

=head1 NAME

Shodo - Auto-generate documents from HTTP::Request and HTTP::Response

=head1 SYNOPSIS

    use HTTP::Request::Common;
    use HTTP::Response;
    use Shodo;

    my $shodo = Shodo->new();
    my $suzuri = $shodo->new_suzuri('An endpoint method.');

    my $req = POST '/entry', [ id => 1, message => 'Hello Shodo' ];
    $suzuri->request($req);
    my $res = HTTP::Response->new(200);
    $res->content('{ "message" : "success" }');
    $suzuri->response($res);

    print $suzuri->document(); # print document as Markdown format

=head1 DESCRIPTION

Shodo generates Web API documents as Markdown format automatically and validates parameters using HTTP::Request/Response.

B<THIS IS A DEVELOPMENT RELEASE. API MAY CHANGE WITHOUT NOTICE>.

=head1 Methods

=head2 new

    my $shodo = Shodo->new(
        document_root => 'doc'
    );

Create and return new Shodo object. "document_root" is optional parameter for your document root directory.

=head2 template

    $shodo->template($tmpl);

Set custom template.

=head2 document_root

    $shodo->document_root('doc');

Set document root directory.

=head2 new_suzuri

    my $suzuri = $shodo->new_suzuri('This is description.');

Create and return new L<Shodo::Suzuri> object with the description.

=head2 stock

    $shodo->stock($suzuri->doc());

Stock text of documents for writing later. The parameter document is anything ok, but Markdown based is recommended.

=head2 write

    $shodo->write('output.md');

Write the documentation in stocks to the file and make the stock empty.

=head1 SEE ALSO

L<Test::Shodo::JSONRPC>

"autodoc": L<https://github.com/r7kamura/autodoc>

L<Test::JsonAPI::Autodoc>

What is Shodo?: L<http://en.wikipedia.org/wiki/Shodo>

=head1 THANKS

Songmu for naming as "Shodo". It's pretty.

Moznion for making Test::JsonAPI::Autodoc.

Hachioji.pm for advising.

=head1 LICENSE

Copyright (C) Yusuke Wada.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Yusuke Wada E<lt>yusuke@kamawada.comE<gt>

=cut
